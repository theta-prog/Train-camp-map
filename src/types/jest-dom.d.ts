/// <reference types="jest" />
/// <reference types="@testing-library/jest-dom" />

declare global {
  namespace jest {
    interface Matchers<R> {
      toBeInTheDocument(): R
      toHaveClass(className: string, ...classNames: string[]): R
      toHaveAttribute(attr: string, value?: string): R
      toHaveTextContent(text: string | RegExp): R
      toBeVisible(): R
      toBeDisabled(): R
      toBeEnabled(): R
      toHaveValue(value: string | number): R
    }
  }
}

export {}
