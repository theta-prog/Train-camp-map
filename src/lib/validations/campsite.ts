import { z } from 'zod'

// キャンプ場登録フォームのバリデーションスキーマ
export const campsiteSchema = z.object({
  // 基本情報
  name_ja: z.string().min(1, '日本語名は必須です').max(100, '100文字以内で入力してください'),
  name_en: z.string().max(100, '100文字以内で入力してください').optional(),
  
  // 位置情報 (Google Mapで選択可能、手動入力も可能)
  lat: z.number({ invalid_type_error: '緯度は数値で入力してください' })
    .min(-90, '緯度は-90以上である必要があります')
    .max(90, '緯度は90以下である必要があります')
    .optional(),
  lng: z.number({ invalid_type_error: '経度は数値で入力してください' })
    .min(-180, '経度は-180以上である必要があります')
    .max(180, '経度は180以下である必要があります')
    .optional(),
  
  // 住所情報
  address_ja: z.string().min(1, '日本語住所は必須です').max(200, '200文字以内で入力してください'),
  address_en: z.string().max(200, '200文字以内で入力してください').optional(),
  
  // 連絡先
  phone: z.string().regex(/^[\d\-\+\(\)\s]*$/, '有効な電話番号を入力してください').optional().or(z.literal('')),
  website: z.string().url('有効なURLを入力してください').optional().or(z.literal('')),
  reservationUrl: z.string().url('有効なURLを入力してください').optional().or(z.literal('')),
  
  // 料金
  price: z.string().min(1, '料金は必須です').max(100, '100文字以内で入力してください'),
  priceMin: z.number().min(0, '最小料金は0以上である必要があります').optional(),
  priceMax: z.number().min(0, '最大料金は0以上である必要があります').optional(),
  
  // チェックイン・チェックアウト
  checkInTime: z.string().max(50, '50文字以内で入力してください').optional().or(z.literal('')),
  checkOutTime: z.string().max(50, '50文字以内で入力してください').optional().or(z.literal('')),
  cancellationPolicy: z.string().max(500, '500文字以内で入力してください').optional().or(z.literal('')),
  
  // アクセス情報
  nearest_station_ja: z.string().min(1, '最寄り駅（日本語）は必須です').max(100, '100文字以内で入力してください'),
  nearest_station_en: z.string().max(100, '100文字以内で入力してください').optional(),
  access_time_ja: z.string().min(1, 'アクセス時間（日本語）は必須です').max(100, '100文字以内で入力してください'),
  access_time_en: z.string().max(100, '100文字以内で入力してください').optional(),
  
  // 説明
  description_ja: z.string().min(1, '説明（日本語）は必須です').max(1000, '1000文字以内で入力してください'),
  description_en: z.string().max(1000, '1000文字以内で入力してください').optional(),
  
  // 設備・アクティビティ
  facilities: z.array(z.string()),
  activities: z.array(z.string()),
  
  // 画像
  images: z.array(z.string().url('有効なURLを入力してください')).optional(),
})

export type CampsiteFormData = z.infer<typeof campsiteSchema>

// 利用可能な設備一覧
export const AVAILABLE_FACILITIES = [
  { id: 'restroom', label_ja: 'トイレ', label_en: 'Restroom' },
  { id: 'shower', label_ja: 'シャワー', label_en: 'Shower' },
  { id: 'parking', label_ja: '駐車場', label_en: 'Parking' },
  { id: 'wifi', label_ja: 'WiFi', label_en: 'WiFi' },
  { id: 'bbq', label_ja: 'BBQ場', label_en: 'BBQ Area' },
  { id: 'kitchen', label_ja: 'キッチン', label_en: 'Kitchen' },
  { id: 'laundry', label_ja: 'ランドリー', label_en: 'Laundry' },
  { id: 'store', label_ja: '売店', label_en: 'Store' },
] as const

// 利用可能なアクティビティ一覧
export const AVAILABLE_ACTIVITIES = [
  { id: 'hiking', label_ja: 'ハイキング', label_en: 'Hiking' },
  { id: 'fishing', label_ja: '釣り', label_en: 'Fishing' },
  { id: 'swimming', label_ja: '水泳', label_en: 'Swimming' },
  { id: 'cycling', label_ja: 'サイクリング', label_en: 'Cycling' },
  { id: 'boating', label_ja: 'ボート', label_en: 'Boating' },
  { id: 'stargazing', label_ja: '星空観察', label_en: 'Stargazing' },
  { id: 'photography', label_ja: '写真撮影', label_en: 'Photography' },
  { id: 'birdwatching', label_ja: 'バードウォッチング', label_en: 'Bird Watching' },
] as const
