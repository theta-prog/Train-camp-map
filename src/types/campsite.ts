export interface Campsite {
  id: string
  name: {
    ja: string
    en: string
  }
  lat: number
  lng: number
  address: {
    ja: string
    en: string
  }
  phone: string
  website: string
  reservationUrl?: string // 予約サイトURL
  price: string // 表示用の価格文字列（例: "¥2,000-¥5,000/泊" or "¥3,000/泊"）
  priceMin?: number // 最小料金（数値）
  priceMax?: number // 最大料金（数値）
  checkInTime?: string // チェックイン時間
  checkOutTime?: string // チェックアウト時間
  cancellationPolicy?: string // キャンセルポリシー
  images?: string[] // 画像URL配列
  facilities: string[]
  activities: string[]
  nearestStation: {
    ja: string
    en: string
  }
  accessTime: {
    ja: string
    en: string
  }
  description: {
    ja: string
    en: string
  }
}

export interface SearchFilters {
  keyword: string
  maxPrice: number
  minPrice?: number // 最小料金フィルター
  facilities: string[]
  activities: string[]
}
