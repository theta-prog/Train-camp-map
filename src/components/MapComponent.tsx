import { useCallback, useState } from 'react'
import { Map, AdvancedMarker, InfoWindow } from '@vis.gl/react-google-maps'
import { useTranslations } from 'next-intl'
import { useParams, useRouter } from 'next/navigation'
import { Campsite } from '@/types/campsite'

interface MapComponentProps {
  campsites: Campsite[]
  onCampsiteSelect: (campsite: Campsite | null) => void
  selectedCampsite: Campsite | null
}

const center = {
  lat: 36.2,
  lng: 139.3 // 関東地方全体が見えるように調整
}

export default function MapComponent({ 
  campsites, 
  onCampsiteSelect, 
  selectedCampsite 
}: MapComponentProps) {
  const t = useTranslations()
  const params = useParams()
  const router = useRouter()
  const locale = (params?.locale as 'ja' | 'en') || 'ja'
  const [activeMarker, setActiveMarker] = useState<string | null>(null)

  const handleMarkerClick = useCallback((campsite: Campsite) => {
    setActiveMarker(campsite.id)
    onCampsiteSelect(campsite)
  }, [onCampsiteSelect])

  const handleInfoWindowClose = useCallback(() => {
    setActiveMarker(null)
  }, [])

  return (
    <div style={{ width: '100%', height: '600px' }}>
      <Map
        defaultCenter={center}
        defaultZoom={8}
        mapId={process.env.NEXT_PUBLIC_GOOGLE_MAPS_MAP_ID || null}
        style={{ width: '100%', height: '100%' }}
        disableDefaultUI={false}
        zoomControl={true}
        gestureHandling={'greedy'}
      >
        {campsites.map((campsite) => (
          <AdvancedMarker
            key={campsite.id}
            position={{ lat: campsite.lat, lng: campsite.lng }}
            onClick={() => handleMarkerClick(campsite)}
            title={campsite.name[locale]}
          >
            <div style={{
              width: '30px',
              height: '30px',
              backgroundColor: '#dc2626',
              borderRadius: '50%',
              display: 'flex',
              alignItems: 'center',
              justifyContent: 'center',
              color: 'white',
              fontSize: '16px',
              border: '2px solid white',
              boxShadow: '0 2px 4px rgba(0,0,0,0.3)'
            }}>
              🏕
            </div>
          </AdvancedMarker>
        ))}

        {activeMarker && selectedCampsite && (
          <InfoWindow
            position={{ lat: selectedCampsite.lat, lng: selectedCampsite.lng }}
            onCloseClick={handleInfoWindowClose}
          >
            <div className="p-2 max-w-xs">
              <h3 className="font-bold text-lg mb-1">{selectedCampsite.name[locale]}</h3>
              <p className="text-sm text-gray-600 mb-2">{selectedCampsite.address[locale]}</p>
              <p className="text-sm mb-1">
                <span className="font-semibold">{t('campsiteList.nearestStation')}:</span> {selectedCampsite.nearestStation[locale]}
              </p>
              <p className="text-sm mb-1">
                <span className="font-semibold">{t('campsiteList.access')}:</span> {selectedCampsite.accessTime[locale]}
              </p>
              <p className="text-sm mb-2">
                <span className="font-semibold">{t('campsiteList.price')}:</span> {selectedCampsite.price}
              </p>
              <button
                className="bg-green-600 text-white px-3 py-1 rounded text-sm hover:bg-green-700"
                onClick={() => {
                  router.push(`/${locale}/campsites/${selectedCampsite.id}`)
                }}
              >
                {t('map.viewDetails')}
              </button>
            </div>
          </InfoWindow>
        )}
      </Map>
    </div>
  )
}
