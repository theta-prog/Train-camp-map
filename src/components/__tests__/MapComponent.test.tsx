import { Campsite } from '@/types/campsite'
import '@testing-library/jest-dom'
import { fireEvent, render, screen } from '@testing-library/react'
import MapComponent from '../MapComponent'

// 環境変数のモック
const originalEnv = process.env
beforeAll(() => {
  process.env = {
    ...originalEnv,
    NEXT_PUBLIC_GOOGLE_MAPS_API_KEY: 'test-api-key',
    NEXT_PUBLIC_GOOGLE_MAPS_MAP_ID: 'test-map-id',
  }
})

afterAll(() => {
  process.env = originalEnv
})

// モック設定
jest.mock('next-intl', () => ({
  useTranslations: () => (key: string) => {
    const translations: Record<string, string> = {
      'campsiteList.nearestStation': 'Nearest Station',
      'campsiteList.access': 'Access', 
      'campsiteList.price': 'Price',
      'map.viewDetails': 'View Details',
    }
    return translations[key] || key
  }
}))

const mockUseParams = jest.fn()
const mockPush = jest.fn()

jest.mock('next/navigation', () => ({
  useParams: () => mockUseParams(),
  useRouter: () => ({
    push: mockPush
  })
}))

jest.mock('@vis.gl/react-google-maps', () => ({
  Map: ({ children }: any) => (
    <div data-testid="mock-map" style={{ width: '100%', height: '100%' }}>
      {children}
    </div>
  ),
  AdvancedMarker: ({ children, onClick, title }: any) => (
    <div data-testid="mock-marker" onClick={onClick} title={title}>
      {children}
    </div>
  ),
  InfoWindow: ({ children, onCloseClick }: any) => (
    <div data-testid="mock-info-window">
      <button data-testid="close-button" onClick={onCloseClick}>×</button>
      {children}
    </div>
  ),
}))

const mockCampsites: Campsite[] = [
  {
    id: '1',
    name: { ja: 'テストキャンプ場1', en: 'Test Campsite 1' },
    lat: 35.6762,
    lng: 139.6503,
    address: { ja: '東京都渋谷区', en: 'Shibuya, Tokyo' },
    phone: '03-1234-5678',
    website: 'https://test1.com',
    price: '¥2,000/泊',
    nearestStation: { ja: '渋谷駅', en: 'Shibuya Station' },
    accessTime: { ja: '徒歩15分', en: '15 min walk' },
    description: { ja: '美しい自然に囲まれたキャンプ場', en: 'Beautiful campsite surrounded by nature' },
    facilities: ['restroom', 'shower', 'parking'],
    activities: ['hiking', 'fishing'],
  },
]

describe('MapComponent', () => {
  const mockOnCampsiteSelect = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
    mockUseParams.mockReturnValue({ locale: 'ja' })
  })

  describe('スナップショットテスト', () => {
    it('基本の地図表示', () => {
      const { container } = render(
        <MapComponent
          campsites={mockCampsites}
          onCampsiteSelect={mockOnCampsiteSelect}
          selectedCampsite={null}
        />
      )
      expect(container.firstChild).toMatchSnapshot()
    })

    it('選択されたキャンプ場でInfoWindow表示', () => {
      const { container } = render(
        <MapComponent
          campsites={mockCampsites}
          onCampsiteSelect={mockOnCampsiteSelect}
          selectedCampsite={mockCampsites[0]!}
        />
      )
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('基本機能', () => {
    it('マーカークリック時にコールバック呼び出し', () => {
      render(
        <MapComponent
          campsites={mockCampsites}
          onCampsiteSelect={mockOnCampsiteSelect}
          selectedCampsite={null}
        />
      )

      const markers = screen.getAllByTestId('mock-marker')
      fireEvent.click(markers[0]!)
      
      expect(mockOnCampsiteSelect).toHaveBeenCalledWith(mockCampsites[0])
    })

    it('InfoWindowクローズ機能で状態をリセット', () => {
      // selectedCampsiteを設定してInfoWindowを表示
      const { rerender } = render(
        <MapComponent
          campsites={mockCampsites}
          onCampsiteSelect={mockOnCampsiteSelect}
          selectedCampsite={null}
        />
      )

      // マーカーをクリックしてonCampsiteSelectを呼び出す
      const markers = screen.getAllByTestId('mock-marker')
      fireEvent.click(markers[0]!)
      
      // selectedCampsiteを設定して再レンダリング
      rerender(
        <MapComponent
          campsites={mockCampsites}
          onCampsiteSelect={mockOnCampsiteSelect}
          selectedCampsite={mockCampsites[0]!}
        />
      )
      
      // InfoWindowが表示されている
      expect(screen.getByTestId('mock-info-window')).toBeInTheDocument()
      
      // クローズボタンをクリックしてhandleInfoWindowCloseを呼び出す
      const closeButton = screen.getByTestId('close-button')
      fireEvent.click(closeButton)
      
      // InfoWindowが閉じられる（activeMarkerがnullになる）
      expect(screen.queryByTestId('mock-info-window')).not.toBeInTheDocument()
    })

    it('handleMarkerClickでactiveMarkerを設定', () => {
      render(
        <MapComponent
          campsites={mockCampsites}
          onCampsiteSelect={mockOnCampsiteSelect}
          selectedCampsite={null}
        />
      )

      // マーカーをクリック
      const markers = screen.getAllByTestId('mock-marker')
      fireEvent.click(markers[0]!)
      
      // onCampsiteSelectが呼ばれることを確認
      expect(mockOnCampsiteSelect).toHaveBeenCalledWith(mockCampsites[0])
    })

    it('paramsがnullの場合のデフォルトロケール処理', () => {
      mockUseParams.mockReturnValue(null)
      
      render(
        <MapComponent
          campsites={mockCampsites}
          onCampsiteSelect={mockOnCampsiteSelect}
          selectedCampsite={null}
        />
      )

      // デフォルトの日本語ロケールで正常に表示される
      expect(screen.getByTestId('mock-map')).toBeInTheDocument()
    })

    it('localeが未定義の場合のフォールバック処理', () => {
      mockUseParams.mockReturnValue({ locale: undefined })
      
      render(
        <MapComponent
          campsites={mockCampsites}
          onCampsiteSelect={mockOnCampsiteSelect}
          selectedCampsite={null}
        />
      )

      // デフォルトの日本語ロケールで正常に表示される
      expect(screen.getByTestId('mock-map')).toBeInTheDocument()
    })

    it('InfoWindow内の詳細ボタンがクリック可能', () => {
      // selectedCampsiteを設定してInfoWindowを表示
      render(
        <MapComponent
          campsites={mockCampsites}
          onCampsiteSelect={mockOnCampsiteSelect}
          selectedCampsite={mockCampsites[0]!}
        />
      )

      // activeMarkerを設定するためにマーカーをクリック
      const markers = screen.getAllByTestId('mock-marker')
      fireEvent.click(markers[0]!)
      
      // InfoWindowが表示される
      expect(screen.getByTestId('mock-info-window')).toBeInTheDocument()
      
      // 詳細ボタンをクリック
      const detailsButton = screen.getByText('View Details')
      fireEvent.click(detailsButton)
      
      // ボタンがクリック可能であることを確認（エラーが発生しないことを確認）
      expect(detailsButton).toBeInTheDocument()
    })
  })
})
