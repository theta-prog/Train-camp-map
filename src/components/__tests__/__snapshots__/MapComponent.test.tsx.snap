// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MapComponent スナップショットテスト 基本の地図表示 1`] = `
<div
  style="width: 100%; height: 600px;"
>
  <div
    data-testid="mock-map"
    style="width: 100%; height: 100%;"
  >
    <div
      data-testid="mock-marker"
      title="テストキャンプ場1"
    >
      <div
        style="width: 30px; height: 30px; background-color: rgb(220, 38, 38); border-radius: 50%; display: flex; align-items: center; justify-content: center; color: white; font-size: 16px; border: 2px solid white; box-shadow: 0 2px 4px rgba(0,0,0,0.3);"
      >
        🏕
      </div>
    </div>
  </div>
</div>
`;

exports[`MapComponent スナップショットテスト 選択されたキャンプ場でInfoWindow表示 1`] = `
<div
  style="width: 100%; height: 600px;"
>
  <div
    data-testid="mock-map"
    style="width: 100%; height: 100%;"
  >
    <div
      data-testid="mock-marker"
      title="テストキャンプ場1"
    >
      <div
        style="width: 30px; height: 30px; background-color: rgb(220, 38, 38); border-radius: 50%; display: flex; align-items: center; justify-content: center; color: white; font-size: 16px; border: 2px solid white; box-shadow: 0 2px 4px rgba(0,0,0,0.3);"
      >
        🏕
      </div>
    </div>
  </div>
</div>
`;
